client/asset/btc/btc.go

```diff
@@ -1854,6 +1855,477 @@ func (btc *baseWallet) fundedTx(coins asset.Coins) (*wire.MsgTx, uint64, []outPo
+	if err != nil {
+		return 0, err
+	}
+
+	return fee, nil
+}
+
+// PreAccelerate returns the current average fee rate of the unmined swap
+// initiation and acceleration transactions, and also returns a suggested
+// range that the fee rate should be increased to in order to expedite mining.
+func (btc *baseWallet) PreAccelerate(swapCoins, accelerationCoins []dex.Bytes, changeCoin dex.Bytes, requiredForRemainingSwaps, feeSuggestion uint64) (currentRate uint64, suggestedRange asset.XYRange, err error) {
+	makeError := func(err error) (uint64, asset.XYRange, error) {
+		return 0, asset.XYRange{}, err
+	}
+
+	changeTxHash, changeVout, err := decodeCoinID(changeCoin)
+	if err != nil {
+		return makeError(err)
+	}
+
+	can, err := btc.changeCanBeAccelerated(changeTxHash, changeVout, requiredForRemainingSwaps)
+	if err != nil {
+		return makeError(err)
+	}
+	if !can {
+		return makeError(errors.New("change cannot be accelerated"))
+	}
+
+	sortedTxChain, err := btc.sortedTxChain(swapCoins, accelerationCoins, changeCoin)
+	if err != nil {
+		return makeError(fmt.Errorf("failed to sort swap chain: %w", err))
+	}
+
+	var swapTxsSize, feesAlreadyPaid uint64
+	for _, tx := range sortedTxChain {
+		if tx.Confirmations > 0 {
+			continue
+		}
+		feesAlreadyPaid += toSatoshi(math.Abs(tx.Fee))
+		msgTx, err := msgTxFromBytes(tx.Hex)
+		if err != nil {
+			return makeError(err)
+		}
+		swapTxsSize += dexbtc.MsgTxVBytes(msgTx)
+	}
+
+	btc.fundingMtx.RLock()
+	utxos, _, utxosVal, err := btc.spendableUTXOs(1)
+	btc.fundingMtx.RUnlock()
+	if err != nil {
+		return makeError(err)
+	}
+
+	var coins asset.Coins
+	for _, utxo := range utxos {
+		coins = append(coins, newOutput(utxo.txHash, utxo.vout, utxo.amount))
+	}
+
+	changeOutput, err := btc.lookupOutput(changeTxHash, changeVout)
+	if err != nil {
+		return makeError(err)
+	}
+
+	tx, _, _, err := btc.fundedTx(append(coins, changeOutput))
+	if err != nil {
+		return makeError(err)
+	}
+	tx, err = btc.node.signTx(tx)
+	if err != nil {
+		return makeError(err)
+	}
+
+	newChangeTxSize := dexbtc.MsgTxVBytes(tx)
+	if requiredForRemainingSwaps > 0 {
+		if btc.segwit {
+			newChangeTxSize += dexbtc.P2WPKHOutputSize
+		} else {
+			newChangeTxSize += dexbtc.P2PKHOutputSize
+		}
+	}
+
+	maxRate := (changeOutput.value + feesAlreadyPaid + utxosVal - requiredForRemainingSwaps) / (newChangeTxSize + swapTxsSize)
+	currentRate = feesAlreadyPaid / swapTxsSize
+
+	if maxRate <= currentRate {
+		return makeError(fmt.Errorf("cannot accelerate, max rate %v <= current rate %v", maxRate, currentRate))
+	}
+
+	maxSuggestion := currentRate * 5
+	if feeSuggestion*5 > maxSuggestion {
+		maxSuggestion = feeSuggestion * 5
+	}
+	if maxRate < maxSuggestion {
+		maxSuggestion = maxRate
+	}
+
+	suggestedRange = asset.XYRange{
+		Start: asset.XYRangePoint{
+			Label: "Min",
+			X:     float64(currentRate+1) / float64(currentRate),
+			Y:     float64(currentRate + 1),
+		},
```

Comment:
The X and the Y are the inputs to the range selector UI element. X is how many times the current rate the selected rate is, and Y is the selected rate. The plus one is because the minimum they should be able to accelerate is 1 satoshi / byte more than the current rate. Of course this is probably still too low, but that's the absolute minimum where you would still be doing some acceleration.

Is generally the design of the dex supposed to be to give users maximum flexibility/responsibility, or hand holding them? If the fee estimation is incorrect, I wouldn't want to force them to pay a higher fee than they feel is necessary.

> Also, are we sure that `feesAlreadyPaid` isn't ever zero, causing a panic here?

I'll update it to make sure there's no divide by 0 issues.

Reply:
Well, it's a design question. So, @chappjc can decide :D